esp-wifi/src/esp_now/mod.rs

```diff
@@ -455,26 +407,56 @@ impl<'d> EspNow<'d> {
     pub fn set_rate(&self, rate: WifiPhyRate) -> Result<(), EspNowError> {
         check_error!({ esp_wifi_config_espnow_rate(wifi_interface_t_WIFI_IF_STA, rate as u32,) })
     }
+}
+
+pub struct EspNowSender<'d>(&'d ());
 
+impl<'d> EspNowSender<'d> {
     /// Send data to peer
     ///
     /// The peer needs to be added to the peer list first.
-    ///
-    /// This method returns a `SendWaiter` on success. ESP-NOW protocol provides guaranteed
-    /// delivery on MAC layer. If you need this guarantee, call `wait` method of the returned
-    /// `SendWaiter` and make sure it returns `SendStatus::Success`.
-    /// However, this method will block current task for milliseconds.
-    /// So you can just drop the waiter if you want high frequency sending.
-    pub fn send(&self, dst_addr: &[u8; 6], data: &[u8]) -> Result<SendWaiter, EspNowError> {
-        let mut addr = [0u8; 6];
-        addr.copy_from_slice(dst_addr);
-
+    pub fn send<'s>(&'s mut self, dst_addr: &[u8; 6], data: &[u8]) -> Result<SendWaiter<'s>, EspNowError> {
         ESP_NOW_SEND_CB_INVOKED.store(false, Ordering::Release);
-        check_error!({ esp_now_send(addr.as_ptr(), data.as_ptr(), data.len() as u32) })?;
-        Ok(SendWaiter(()))
+        check_error!({ esp_now_send(dst_addr.as_ptr(), data.as_ptr(), data.len() as u32) })?;
+        Ok(SendWaiter(PhantomData))
+    }
+}
+
+/// This struct is returned by a sync esp now send. Invoking `wait` method of this
+/// struct will block current task until the callback function of esp now send is called
+/// and return the status of previous sending. 
+/// 
+/// This waiter borrows the sender, so when used in multiple tasks, the lock will only be
+/// released when the waiter is dropped or consumed via `wait`
+pub struct SendWaiter<'s>(PhantomData<&'s mut EspNowSender<'s>>);
+
+impl<'s> SendWaiter<'s> {
+    /// Wait for the previous sending to complete, i.e. the send callback is invoked with 
+    /// status of the sending.
+    pub fn wait(self) -> Result<(), EspNowError> {
+        // prevent redundant waiting since we waits for the callback in the Drop implementation
+        core::mem::forget(self);
+        while !ESP_NOW_SEND_CB_INVOKED.load(Ordering::Acquire) {}
+
+        if ESP_NOW_SEND_STATUS.load(Ordering::Relaxed) {
+            Ok(())
+        } else {
+            Err(EspNowError::SendFailed)
+        }
     }
+}
 
-    /// Receive data
+impl<'s> Drop for SendWaiter<'s> {
+    /// wait for the send to complete to prevent the lock on `EspNowSender` get unlocked 
+    /// before a callback is invoked.
+    fn drop(&mut self) {
+        while !ESP_NOW_SEND_CB_INVOKED.load(Ordering::Acquire) {}
+    }
+}
```

Comment:
Would enforcing waiting the callback in the drop trait a good idea? Or we can just suggest users to call the `wait` method to avoid getting wrong status.

Reply:
I _think_ waiting for the callback is not too bad. Will be in an order of tens of milliseconds and users like to skip reading documentation

Reply:
Currently my async implementation of `send` invokes `esp_now_send` on the first poll and then waits for the callback. So user can bypass the waiting part by only polling once. Personally I think its okay since this is pretty rare. I'd like to hear your opinion.

Reply:
I guess that's fine. Maybe worth to document the behavior in the API docs for the function

Reply:
Ok. Thank you for your suggestion.